### ISSUE TITLE
UnicodeDecodeError: 'utf-8' codec can't decode byte

### ISSUE THREAD

**Issue by marty-optum** [microsoft/IoT-For-Beginners#354]
I am going through the [Decode GPS data](https://github.com/microsoft/IoT-For-Beginners/blob/main/3-transport/lessons/1-location-tracking/single-board-computer-gps-decode.md#decode-gps-data---virtual-iot-hardware-and-raspberry-pi) lesson using a Raspberry Pi. I am using VS Code 1.63.2 to edit remotely from a mac with macOS 11.6 (Big Sur).  The Raspberry Pi is using Python 3.7.3.

I have successfully completed the task to [connect a GPS sensor and read GPS data](https://github.com/microsoft/IoT-For-Beginners/blob/main/3-transport/lessons/1-location-tracking/pi-gps-sensor.md), so I know that the GPS sensor is receiving data.

However, when I try to run the program to decode the data, I get errors.  These errors occur even if I am using the sample code copied right from the [code-gps-decode/pi](https://github.com/microsoft/IoT-For-Beginners/blob/main/3-transport/lessons/1-location-tracking/code-gps-decode/pi/gps-sensor/app.py) folder in the lesson's repo.  It also occurs regardless of whether I am running the code from the VS Code built in terminal, or a separate ssh session using macOS terminal.

The error message return is as follows:

```
Traceback (most recent call last):
  File "app2.py", line 25, in <module>
    line = serial.readline().decode('utf-8')
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xa9 in position 2: invalid start byte
```

Note: the values for _byte_ and _position_ in the error message will vary between different attempts.

Here is a screenshot with further detail:

![image](https://user-images.githubusercontent.com/71525323/147115770-b81cb50e-678c-469a-80eb-dcaa5b3b8294.png)

From what I can guess, it looks like the data being read from the GPS sensor includes different characters at different positions which were not encoded using **utf-8**, or there is something else wrong with the `decode` function in the `pyserial` library.

@jimbobbennett , full disclosure: I am not a Python expert, so it's entirely possible that I could be doing something wrong, and that this is a (hopefully) easy fix.  I do think that it is something that other students will come across, however, so hopefully you provide some guidance.  If you need more information or need me to try some other tests, just let me know.

[UPDATE 12/29]:  After further testing, I can get the code to run after trying **multiple** (sometimes 12 or 13) times.  It looks to me like the `serial.readline().decode('utf-8')` method will sometimes get a partial message as its first line, then chokes because it doesn't know how to deal with an incomplete line/format.  @jimbobbennett, is there a way to modify the sample code in this lesson to validate that the first line read from the GPS serial interface is a complete, well-formatted message _before_ trying to call the decode method?  To me, the lesson should yield code that produces a consistent output/experience each time, otherwise how does the student know if the error is coming from a bad library, or something that they may have done wrong?

**Comment by jimbobbennett** [comment]
Good catch here - I need to tidy the code up a bit.

Essentially what is happening is the serial connection sends data a byte at a time, but UTF-8 contains variable-width characters. This means you can read a byte that is part way through a character, which will fail to decode. This will only happen the first time, as after that full lines will be read.

It runs sometimes because of the random chance that the byte being read is not part way through a character. The code needs to be tweaked to be something like this:

```python
try:
    line = serial_conn.readline().decode('utf-8')
 except UnicodeDecodeError:
    line = serial_conn.readline().decode('utf-8')
```

This will read a line, and if the line starts part-way through a character, it catches the exception and reads the next full line. This should work as each line will be read correctly.

**Comment by jimbobbennett** [comment]
Let me make some fixes to the code/text to sort this

**Comment by marty-optum** [comment]
@jimbobbennett ,  I hope you had a restful holiday!  I tested the snippet of code you provided above and it has fixed the problem I was seeing for 3 out of 3 attempts!

Note the following:

- the `serial_conn` method did not resolve for me, so I stayed with `serial`
- I replaced the `line = serial.readline().decode('utf-8')` line in **both** occurrences (before and within the `while` loop)
- I bumped out the `except UnicodeDecodeError:` line **one character to the left** so that it vertically aligned with the `try:` clause above it

**Comment by jimbobbennett** [comment]
Good catches on my bad copy/paste from another repo!

**Comment by maujac** [comment]
Thank you @marty-optum for opening the issue and @jimbobbennett for the proposed solution. It worked like a charm for me.

Is this repo open for contributions? I've created PR #374 to add this solution as a hint in the [Read GPS data - Raspberry Pi](https://github.com/microsoft/IoT-For-Beginners/blob/main/3-transport/lessons/1-location-tracking/pi-gps-sensor.md#read-gps-data---raspberry-pi) lesson.

**Comment by jimbobbennett** [comment]
Fixed by #374, thanks @maujac.

**Comment by ismaelmesa00** [comment]
Hi, I am with a python project for a 3D printing and I am having a similar issue that I can't repair even doing what you suggest.
My problem is that when I initially run the code everything goes perfect, however after 15 minutes of doing the process the error apears. I try change the decode from UTF-8 to cp1252, ascii and others without luck. Besides due to is a long programm is dificult make long changes. I adjunt a picture of the error, the line of the code and all the scrip. I hope you can help me because I am desperate.
[problem.zip](https://github.com/microsoft/IoT-For-Beginners/files/8868784/problem.zip)
